src/library/trackloader.h

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include <QObject>
+#include <QPointer>
+
+#include "track/track.h"
+#include "track/trackref.h"
+
+class TrackCollectionManager;
+
+namespace mixxx {
+
+// Loads tracks thread-safe and asynchronously from the local collection.
+// Must be located in the same thread where TrackCollectionManager lives!
+class TrackLoader: public QObject {
+    Q_OBJECT
+
+  public:
+    explicit TrackLoader(
+            TrackCollectionManager* trackCollectionManager,
+            QObject* parent = nullptr);
+    ~TrackLoader() override = default;
+
+    // Thread-safe, possibly asynchronous invocation. May emit
+    // the trackLoaded signal directly when invoked from the
+    // same thread!
+    void invokeLoadTrack(
```

Comment:
It is not immediately clear to me when calling code should use this function or `loadTrack` because they are both public. Do they both need to be public? Maybe some comments could clarify? Or maybe we can think of a more self-explaining name for `invokeLoadTrack`?

Reply:
The function `invokeLoadTrack` can be called directly from any thread. It is just a convenience function for emitting the corresponding signal `loadTrack`. Or should we name it `emitLoadTrack`?

Reply:
Even thought `loadTrack()` is a public slot and function it is not allowed to be called from arbitrary threads! That's a deficiency of the C++ type system and the way how Qt implements slots. Qt should check that the slot is only invoked from the object's thread.

Reply:
Okay, please document that the slot is not thread safe.

Reply:
Does `loadTrack` actually have to be public? If so, please document that it should not be invoked by calling code. Developers looking at this code should not get confused which function to use.

Maybe rename `loadTrack` to `loadTrackInternal` and rename `invokeLoadTrack` to `loadTrack`?

Reply:
We should not document basic Qt principles. No slot is thread-safe, i.e. each slot must only be invoked from the QObject's event loop thread! Otherwise we would need to document each and every slot.

Reply:
> Maybe rename loadTrack to loadTrackInternal and rename invokeLoadTrack to loadTrack?

or rename `loadTrack` to `slotLoadTrack` and `invokeLoadTrack` to `loadTrack`?

Reply:
I think the comment here in the header could be clarified: "Thread safe, possibly asynchronous wrapper around slotLoadTrack. May emit the trackLoaded signal directly when invoked from same thread."

Reply:
The new comment is good enough, although I think the language of a "wrapper" around the other function would improve it further.

Reply:
`loadTrack()` is a perfectly valid public slot, no need to rename it as "internal". You can either invoke the track loading functionality by emitting a signal connected to this slot (letting Qt take care of the thread-safety implicitly) or by calling the thread-safe public utility function.

I'm not aware of any conventions in Qt how to handle this use case properly?

Reply:
Conventionally in other parts of Mixxx we prefix slot function names with `slot`. Are we moving away from that convention? I don't have a strong opinion either way.

Reply:
I think it would help if the function name of `invokeLoadTrack` made it explicit that it is invoking/wrapping a slot. How about renaming it to `invokeSlotLoadTrack`?

Reply:
The naming of signals and slots in Mixxx is inconsistent and Qt does not propose any naming conventions.

Signals communicate an event that has happened and should end with a past-participle verb. On the other hand slots are sometimes but not always commands as in this case for which the imperative form is recommended. For slots that are event handlers I usually prefer *onSomething*.

See also: https://cqrs.nu/Faq/commands-and-events

Reply:
This is a very simple use case that allows to discuss some options ;)

The *aoide* gateway exposes it's functions in a similar way. All network requests are strictly asynchronous independent of the caller thread.

Reply:
After we have sorted out the naming here I will align the reusable, asychronous web client.

Reply:
I think the names `invokeSlotLoadTrack`/`slotLoadTrack` would clarify the relationship between the functions. One is a slot, the other calls the slot.